src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationTaggerProvider.cs

```diff
@@ -38,45 +56,135 @@ internal partial class SemanticClassificationTaggerProvider : AbstractAsynchrono
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            await ProduceTagsAsync(context, spanToTag).ConfigureAwait(false);
         }
 
-        protected override ProducerPopulatedTagSource<IClassificationTag> CreateTagSourceCore(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        private async Task ProduceTagsAsync(TaggerContext<IClassificationTag> context, DocumentSnapshotSpan spanToTag)
         {
-            return new SemanticBufferTagSource<IClassificationTag>(
-                subjectBuffer,
-                CreateTagProducer(),
-                CreateEventSource(textViewOpt, subjectBuffer),
-                AsyncListener,
-                NotificationService,
-                this.RemoveTagsThatIntersectEdits,
-                this.SpanTrackingMode);
+            if (await TryClassifyContainingMemberSpan(context, spanToTag).ConfigureAwait(false))
+            {
+                return;
+            }
+
+            // We weren't able to use our specialized codepaths for semantic classifying. 
+            // Fall back to classifying the full span that was asked for.
+            await ClassifySpansAsync(context, spanToTag).ConfigureAwait(false);
         }
 
-        protected override ITagProducer<IClassificationTag> CreateTagProducer()
+        private async Task<bool> TryClassifyContainingMemberSpan(TaggerContext<IClassificationTag> context, DocumentSnapshotSpan spanToTag)
         {
-            return new TagProducer(_typeMap);
+            var range = context.TextChangeRange;
+            if (range == null)
+            {
+                // There was no text change range, we can't just reclassify a member body.
+                return false;
+            }
+
+            // there was top level edit, check whether that edit updated top level element
+            var document = spanToTag.Document;
+            var service = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
+            if (service == null)
+            {
+                return false;
+            }
+
+            // perf optimization. Check whether all edits since the last update has happened within
+            // a member. If it did, it will find the member that contains the changes and only refresh
+            // that member.  If possible, try to get a speculative binder to make things even cheaper.
+
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var changedSpan = new TextSpan(range.Value.Span.Start, range.Value.NewLength);
+            var member = service.GetContainingMemberDeclaration(root, changedSpan.Start);
+            if (member == null || !member.FullSpan.Contains(changedSpan))
+            {
+                // The edit was not fully contained in a member.  Reclassify everything.
```

Comment:
there used to be a top level version check (not sure it is still kept in this change or not) before code gets into this.

so, the case you described above, top level version will be different, and it would shortcut to re-calculating whole file. (by the way, this approach - us calculating whole file semantic classification up-front on every top level type - was one of biggest contributor to our allocation. except diagnostic, semantic classification alone allocates about 20% of all Roslyn allocations and actually never used - so 99% of the calculation is a waste). 

without this optimization, I think Kayle found allocation produced by semantic classification is over 40% of whole Roslyn allocations.